syntax = "proto3";

package telemetry;

service TelemetryService {
  // Stream telemetry data from a switch
  rpc SubscribeTelemetry (TelemetryRequest) returns (stream TelemetryData) {}
}

message TelemetryRequest {
  string switch_id = 1;
  string port = 2;  // Optional, if empty will stream all ports
}

message TelemetryData {
  string timestamp = 1;
  string port = 2;
  string linkState = 3;
  float SNR = 4;
  float FEC_Correctable = 5;
  float FEC_Uncorrectable = 6;
  float CRCErrorCount = 7;
  float Temperature = 8;
  float Voltage = 9;
  float FanSpeed = 10;
  float Humidity = 11;
  float Airflow = 12;
  float AmbientTemperature = 13;
  float OpticalRxPower = 14;
  float OpticalTxPower = 15;
  float LinkLatency = 16;
  float CableLengthEstimate = 17;
  int32 ConnectorInsertionCount = 18;
}